engine/robot.py

```diff
@@ -330,10 +361,63 @@ def set_phase(self, new_phase):
         with open(CSV_PATH + '/phases.csv', 'a') as fd:
             fd.write(str(self.phase) + '\n')
 
-    def execute_avoid_obstacle(self):
-        # TODO: SET BACK TO ORIGINAL MISSION (TRAVERSE OR RETURN)
+    def track_obstacle(self):
         pass
 
+    def execute_avoid_obstacle(self, dist_to_goal, prev_phase):
+        pass
+
+    # Currently, algorithm only supports using the right side sensors of the robot for boundary following.
+    # When encountering an obstacle, robot will begin boundary following by going clockwise around the object.
+    # This was arbitrarily decided since currently there is no better way to determine which direction the robot should
+    # navigate around the obstacle.
+    def execute_boundary_following(self, min_dist):
+        front_dist = self.rf_ultrasonic.distance()
+        back_dist = self.rb_ultrasonic.distance()
+        margin = 1 
+        # Plus or minus distance value used to calculate if robot is parallel to an non-uniform object 
+        # (i.e. not a flat surface). forwardRightSensorReading - backRightSensorReading < margin means
+        # robot is parallel to object.
+        forward_dist = 0.01  # Distance moved forward by robot in one iteration of this method
+        turn_angle = math.pi / 90
+        # Turn angle of robot in one iteration of this method. Robot turns turn_angle if
+        # forwardRightSensorReading > backRightSensorReading and -turn_angle if
+        # forwardRightSensorReading < backRightSensorReading
+
+        if math.abs(front_dist - back_dist) < margin:
+            self.move_forward(forward_dist)
+        elif front_dist > back_dist:
+            self.turn(turn_angle)
+        else:
+            self.turn(-1 * turn_angle)
+
+        '''
+        Robot will always begin boundary following by going clockwise (turning left when encountering an obstacle)
+        1. Determine if robot is parallel using the ultrasonic sensor data
+            * Three cases:
+                1. Front right sensor is farther away from the obstacle than the back right sensor
+                    * Turn right until sensor inputs are equalized (front of the robot will turn closer toward the obstacle)
+                2. Front right sensor is closer to the obstacle than the back right sensor
+                    *  Turn left until sensor inputs are equalized (front of the robot will turn away from the obstacle)
+                3. Both right-side sensors displaying the same value
+                    * With a margin of error accounting for not uniform obstacles
+        2. Once both sensors are displaying the same values, move forward (boundary following)
+        3. Exits boundary following when conditions met in execute_avoid_obstacle() method
+        '''
```

Comment:
Move this specification to directly under the function header? + merge with the comments above it?